File: shreerakhi_customizations/shree/report/customer_item_catalogue/customer_item_catalogue.py
import frappe
from frappe.utils import get_url
from frappe.utils.pdf import get_pdf
from frappe import _
import json

def execute(filters=None):
    filters = filters or {}
    columns = [
        {"label": "Image Link", "fieldname": "image_link", "fieldtype": "HTML", "width": 150},
        {"label": "Item Code", "fieldname": "item_code", "fieldtype": "Link", "options": "Item", "width": 100},
        {"label": "Item Name", "fieldname": "item_name", "fieldtype": "Data", "width": 100},
        {"label": "Item Group", "fieldname": "item_group", "fieldtype": "Link", "options": "Item Group", "width": 100},
        {"label": "Item Range", "fieldname": "item_range", "fieldtype": "Data", "width": 100},
        {"label": "Item Range Name", "fieldname": "item_range_name", "fieldtype": "Data", "width": 100},
        {"label": "Box Type", "fieldname": "box_type", "fieldtype": "Data", "width": 100},
        {"label": "Item Category", "fieldname": "item_category", "fieldtype": "Data", "width": 100},
        {"label": "Available Qty", "fieldname": "available_qty", "fieldtype": "Int", "width": 100},
        {"label": "Stock UOM", "fieldname": "stock_uom", "fieldtype": "Link", "options": "UOM", "width": 100},
        {"label": "Selling Price", "fieldname": "selling_price", "fieldtype": "Currency", "width": 100}
    ]

    # Get filter values
    price_list = filters.get("price_list") or "Standard Selling"
    item_group_filter = filters.get("item_group_filter") or ""
    min_qty = filters.get("min_qty")
    item_categories = filters.get("item_categories") or []
    
    # Handle item_categories if it's a string (from URL params)
    if isinstance(item_categories, str):
        try:
            item_categories = json.loads(item_categories)
        except:
            item_categories = [cat.strip() for cat in item_categories.split(',') if cat.strip()]
    
    base_url = get_url()
    data = []

    # Build the base SQL query
    sql_query = """
        SELECT 
            i.image AS image,
            i.item_code,
            i.item_name,
            i.item_group,
            IFNULL(i.custom_item_range, '') AS item_range,
            IFNULL(i.custom_item_range_name, '') AS item_range_name,
            IFNULL(i.custom_box_type, '') AS box_type,
            IFNULL(i.custom_item_category, '') AS item_category,
            (bin.actual_qty 
            - IFNULL((SELECT SUM(si_item.qty) 
                      FROM `tabSales Invoice Item` si_item
                      JOIN `tabSales Invoice` si ON si.name = si_item.parent
                      WHERE si_item.item_code = i.item_code 
                        AND si.docstatus = 0), 0)
            ) AS available_qty,
            i.stock_uom,
            (SELECT ip.price_list_rate 
             FROM `tabItem Price` ip
             WHERE ip.item_code = i.item_code
               AND ip.price_list = %(price_list)s
             ORDER BY ip.valid_from DESC
             LIMIT 1) AS selling_price
        FROM 
            `tabItem` i
        JOIN 
            `tabBin` bin ON bin.item_code = i.item_code
        WHERE 1=1
    """
    
    # Build query parameters
    query_params = {
        "price_list": price_list
    }
    
    # Add item group filter only if specified
    if item_group_filter:
        sql_query += " AND i.item_group LIKE %(item_group_pattern)s"
        query_params["item_group_pattern"] = f"{item_group_filter}%"
    
    # Add minimum quantity filter only if specified
    if min_qty is not None and min_qty > 0:
        sql_query += """ AND (bin.actual_qty - IFNULL((SELECT SUM(si_item.qty) 
                                          FROM `tabSales Invoice Item` si_item
                                          JOIN `tabSales Invoice` si ON si.name = si_item.parent
                                          WHERE si_item.item_code = i.item_code 
                                            AND si.docstatus = 0),0)) > %(min_qty)s"""
        query_params["min_qty"] = min_qty
    
    # Add category filter only if categories are selected
    if item_categories and len(item_categories) > 0:
        # Create placeholders for categories
        category_placeholders = ", ".join([f"%(category_{i})s" for i in range(len(item_categories))])
        sql_query += f" AND i.custom_item_category IN ({category_placeholders})"
        
        # Add categories to query params
        for i, category in enumerate(item_categories):
            query_params[f"category_{i}"] = category

    try:
        data = frappe.db.sql(sql_query, query_params, as_dict=1) or []
    except Exception as e:
        frappe.log_error(f"Customer Item Catalogue query failed: {e}")

    # Process data
    for row in data:
        # Convert Available Qty to integer
        if row.get("available_qty") is not None:
            row["available_qty"] = int(row["available_qty"])

        img = row.get("image")
        if img:
            if not img.startswith("http"):
                img = f"{base_url}{img}"

            # PDF render
            row["image_html"] = f"""
                <img src="{img}" 
                     style="width:120px; height:120px; object-fit:contain; border:1px solid #ddd; border-radius:6px;">
            """

            # ERPNext report → clickable link
            row["image_link"] = f'<a href="{img}" target="_blank">{img}</a>'
        else:
            row["image_html"] = """
                <div style="width:120px; height:120px; border:1px solid #eee; background:#fafafa;"></div>
            """
            row["image_link"] = "No Image"

    return columns, data

@frappe.whitelist()
def get_item_categories():
    """Get all unique item categories for the filter"""
    try:
        categories = frappe.db.sql("""
            SELECT DISTINCT custom_item_category 
            FROM `tabItem`
            WHERE custom_item_category IS NOT NULL 
              AND custom_item_category != ''
            ORDER BY custom_item_category
        """, as_dict=1)
        
        return [cat.get('custom_item_category') for cat in categories]
    except Exception as e:
        frappe.log_error(f"Failed to fetch item categories: {e}")
        return []

@frappe.whitelist()
def download_customer_catalogue(price_list=None, item_group_filter=None, min_qty=None, item_categories=None):
    # Convert min_qty to integer if provided
    if min_qty:
        try:
            min_qty = int(min_qty)
        except:
            min_qty = None
    
    # Handle item_categories parameter
    if item_categories:
        if isinstance(item_categories, str):
            try:
                item_categories = json.loads(item_categories)
            except:
                item_categories = [cat.strip() for cat in item_categories.split(',') if cat.strip()]
    
    columns, data = execute(filters={
        "price_list": price_list,
        "item_group_filter": item_group_filter,
        "min_qty": min_qty,
        "item_categories": item_categories
    })

    base_url = get_url()
    context = {
        "items": data,
        "price_list": price_list,
        "item_group_filter": item_group_filter,
        "item_categories": item_categories,
        "base_url": base_url,
    }

    html = frappe.render_template(
        "shreerakhi_customizations/templates/includes/customer_catalogue_template.html",
        context
    )

    try:
        pdf_content = get_pdf(html)
    except Exception:
        frappe.throw(_("PDF generation failed (check image links or wkhtmltopdf setup)."))

    frappe.local.response.filename = "Customer_Catalogue.pdf"
    frappe.local.response.filecontent = pdf_content
    frappe.local.response.type = "download"

File: shreerakhi_customizations/shree/report/customer_item_catalogue/customer_item_catalogue.js
frappe.query_reports["Customer Item Catalogue"] = {
    filters: [
        {
            fieldname: "price_list",
            label: __("Price List"),
            fieldtype: "Link",
            options: "Price List",
            default: "Standard Selling"
        },
        {
            fieldname: "item_group_filter",
            label: __("Item Group Filter"),
            fieldtype: "Link",
            options: "Item Group",
            default: ""
        },
        {
            fieldname: "min_qty",
            label: __("Minimum Available Qty"),
            fieldtype: "Int",
            default: 0
        },
        {
            fieldname: "item_categories",
            label: __("Item Categories"),
            fieldtype: "MultiSelectList",
            get_data: function(txt) {
                return frappe.db.get_list('Item Category', {
                    fields: ['name'],
                    filters: {
                        name: ['like', '%' + txt + '%']
                    },
                    order_by: 'name'
                }).then(r => {
                    return r.map(d => d.name);
                });
            }
        }
    ],
    
    onload: function(report) {
        report.page.add_inner_button(__("Download Catalogue PDF"), function() {
            let filters = report.get_values();
            
            // Check if report has data
            if (!report.data || report.data.length === 0) {
                frappe.msgprint({
                    title: __('No Data'),
                    indicator: 'red',
                    message: __('Please run the report first to see items.')
                });
                return;
            }
            
            // Show item selection dialog
            show_item_selection_dialog(report, filters);
        });
        
        function show_item_selection_dialog(report, filters) {
            // Prepare items list with details
            let items_html = `
                <div style="margin-bottom: 15px; display: flex; justify-content: space-between; align-items: center;">
                    <div>
                        <button class="btn btn-xs btn-primary" id="select-all-items">Select All</button>
                        <button class="btn btn-xs btn-default" id="deselect-all-items">Deselect All</button>
                    </div>
                    <span style="font-weight: bold;" id="selected-count">
                        Selected: ${report.data.length} of ${report.data.length}
                    </span>
                </div>
                <div style="max-height: 500px; overflow-y: auto; border: 1px solid #d1d8dd; padding: 10px; background: #fafbfc;">
            `;
            
            report.data.forEach(function(row, index) {
                // Get image URL
                let image_url = '';
                if (row.image) {
                    // Extract image URL from the HTML anchor tag
                    let temp_div = document.createElement('div');
                    temp_div.innerHTML = row.image_link;
                    let anchor = temp_div.querySelector('a');
                    image_url = anchor ? anchor.href : '';
                }
                
                items_html += `
                    <div style="padding: 10px; border-bottom: 1px solid #e8e8e8; background: white; margin-bottom: 8px; border-radius: 4px;">
                        <label style="margin: 0; cursor: pointer; display: flex; align-items: center; gap: 12px;">
                            <input type="checkbox" 
                                   class="item-checkbox" 
                                   data-item-code="${row.item_code}"
                                   data-index="${index}"
                                   checked
                                   style="margin: 0; flex-shrink: 0;">
                            
                            ${image_url ? `
                                <img src="${image_url}" 
                                     style="width: 60px; height: 60px; object-fit: cover; border: 1px solid #ddd; border-radius: 4px; flex-shrink: 0;"
                                     onerror="this.style.display='none'">
                            ` : `
                                <div style="width: 60px; height: 60px; background: #f0f0f0; border: 1px solid #ddd; border-radius: 4px; display: flex; align-items: center; justify-content: center; flex-shrink: 0; font-size: 10px; color: #999;">
                                    No Image
                                </div>
                            `}
                            
                            <div style="flex: 1; min-width: 0;">
                                <div style="font-weight: 600; font-size: 14px; color: #2c3e50; margin-bottom: 4px;">
                                    ${row.item_code}
                                </div>
                                <div style="font-size: 13px; color: #5a5a5a; margin-bottom: 2px;">
                                    ${row.item_name || 'N/A'}
                                </div>
                                ${row.item_group ? `
                                    <div style="font-size: 12px; color: #8D99A6;">
                                        <i class="fa fa-folder-o"></i> ${row.item_group}
                                    </div>
                                ` : ''}
                            </div>
                            
                            <div style="text-align: right; flex-shrink: 0; padding-left: 10px;">
                                <div style="font-size: 13px; color: #666; margin-bottom: 4px;">
                                    Available Qty
                                </div>
                                <div style="font-size: 18px; font-weight: bold; color: ${row.available_qty > 0 ? '#27ae60' : '#e74c3c'};">
                                    ${row.available_qty !== undefined ? row.available_qty : 'N/A'}
                                </div>
                                ${row.stock_uom ? `
                                    <div style="font-size: 11px; color: #999;">
                                        ${row.stock_uom}
                                    </div>
                                ` : ''}
                            </div>
                        </label>
                    </div>
                `;
            });
            
            items_html += '</div>';
            
            // Create dialog
            let d = new frappe.ui.Dialog({
                title: __('Select Items for Catalogue'),
                fields: [
                    {
                        fieldtype: 'HTML',
                        fieldname: 'items_list',
                        options: items_html
                    }
                ],
                primary_action_label: __('Download PDF'),
                primary_action: function() {
                    // Get selected items
                    let selected_items = [];
                    d.$wrapper.find('.item-checkbox:checked').each(function() {
                        selected_items.push($(this).data('item-code'));
                    });
                    
                    if (selected_items.length === 0) {
                        frappe.msgprint({
                            title: __('No Items Selected'),
                            indicator: 'orange',
                            message: __('Please select at least one item.')
                        });
                        return;
                    }
                    
                    // Close dialog and download
                    d.hide();
                    download_catalogue(filters, selected_items);
                }
            });
            
            d.show();
            
            // Update counter function
            function update_counter() {
                let total = d.$wrapper.find('.item-checkbox').length;
                let checked = d.$wrapper.find('.item-checkbox:checked').length;
                d.$wrapper.find('#selected-count').text(`Selected: ${checked} of ${total}`);
            }
            
            // Select All button
            d.$wrapper.find('#select-all-items').on('click', function() {
                d.$wrapper.find('.item-checkbox').prop('checked', true);
                update_counter();
            });
            
            // Deselect All button
            d.$wrapper.find('#deselect-all-items').on('click', function() {
                d.$wrapper.find('.item-checkbox').prop('checked', false);
                update_counter();
            });
            
            // Individual checkbox change
            d.$wrapper.find('.item-checkbox').on('change', function() {
                update_counter();
            });
        }
        
        function download_catalogue(filters, selected_items) {
            // Prepare item_categories parameter
            let categories = filters.item_categories || [];
            let categoriesParam = "";
            
            if (categories) {
                let categoriesArray = [];
                
                if (Array.isArray(categories)) {
                    categoriesArray = categories.filter(c => c);
                } else if (typeof categories === 'string' && categories.trim()) {
                    categoriesArray = categories.split(',').map(c => c.trim()).filter(c => c);
                }
                
                if (categoriesArray.length > 0) {
                    categoriesParam = "&item_categories=" + encodeURIComponent(JSON.stringify(categoriesArray));
                }
            }

            // Build URL
            let url_params = [];
            
            url_params.push("price_list=" + encodeURIComponent(filters.price_list || "Standard Selling"));
            
            if (filters.item_group_filter) {
                url_params.push("item_group_filter=" + encodeURIComponent(filters.item_group_filter));
            }
            
            if (filters.min_qty && filters.min_qty > 0) {
                url_params.push("min_qty=" + filters.min_qty);
            }
            
            if (categoriesParam) {
                url_params.push(categoriesParam.substring(1));
            }
            
            // Add selected items
            if (selected_items && selected_items.length > 0) {
                url_params.push("selected_items=" + encodeURIComponent(JSON.stringify(selected_items)));
            }

            // Show loading message
            frappe.show_alert({
                message: __('Generating PDF with {0} items...', [selected_items.length]),
                indicator: 'blue'
            }, 5);

            // Download
            window.open(
                frappe.urllib.get_full_url(
                    "/api/method/shreerakhi_customizations.shree.report.customer_item_catalogue.customer_item_catalogue.download_customer_catalogue"
                    + "?" + url_params.join("&")
                )
            );
        }
    }
};